### PR TITLE
make populateUncoveredIntervals a configuration in query context

### LINKED ISSUE #2108: uncoveredIntervals can overflow query response header
CachingClusteredClient builds up a list of "uncovered intervals", intervals within the query interval that had no data in the underlying segments.

This list is returned as part of the response header in the "X-Druid-Response-Context" field.

If there are a large number of uncovered intervals, the response header may not be large enough to hold this list. On my local system I saw the header buffer had an 8KB size limit.

I noticed this while running a SegmentMetadataQuery on a set of minute-granularity segments that only contained data for even minutes.

e.g.
```
2015-12-16T17:33:21,646 WARN [qtp296974277-44] org.eclipse.jetty.server.HttpChannel - /druid/v2/?pretty
java.io.IOException: Response header too large
	at org.eclipse.jetty.http.HttpGenerator.generateResponse(HttpGenerator.java:400) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpConnection$SendCallback.process(HttpConnection.java:599) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:246) ~[jetty-util-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:208) ~[jetty-util-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpConnection.send(HttpConnection.java:448) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpChannel.sendResponse(HttpChannel.java:762) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpChannel.write(HttpChannel.java:800) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:139) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:132) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:347) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
...
Caused by: java.nio.BufferOverflowException
	at java.nio.Buffer.nextPutIndex(Buffer.java:521) ~[?:1.8.0_51]
	at java.nio.HeapByteBuffer.put(HeapByteBuffer.java:169) ~[?:1.8.0_51]
	at org.eclipse.jetty.http.HttpGenerator.putSanitisedValue(HttpGenerator.java:1057) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.http.HttpGenerator.putTo(HttpGenerator.java:1079) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.http.HttpGenerator.generateHeaders(HttpGenerator.java:703) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.http.HttpGenerator.generateResponse(HttpGenerator.java:385) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	... 74 more
```

### PR DESCRIPTION
**Pull request by himanshug**
fixes #2108

### CODE REVIEW

File: processing/src/main/java/io/druid/query/BaseQuery.java [hunk 1]
```diff
@@ -62,6 +62,11 @@
     return parseBoolean(query, "finalize", defaultValue);
   }
 
+  public static <T> int getContextPopulateUncoveredIntervals(Query<T> query, int defaultValue)
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/BaseQuery.java]: why not boolean?

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/BaseQuery.java]: ah I see. this is a limit not a boolean. can this be `uncoveredIntervalsLimit` or similar?

Comment (himanshug) [inline processing/src/main/java/io/druid/query/BaseQuery.java]: so that user can keep limit the size even if enabled.

Comment (himanshug) [inline processing/src/main/java/io/druid/query/BaseQuery.java]: sure, i can rename the key to `uncoveredIntervalsLimit` if that is more appropriate.

File: server/src/main/java/io/druid/client/CachingClusteredClient.java [hunk 2]
```diff
@@ -166,33 +166,54 @@ public CachingClusteredClient(
...
+        // This returns intervals for which NO segment is present.
+        // Which is not necessarily an indication that the data doesn't exist or is
+        // incomplete. The data could exist and just not be loaded yet.  In either
+        // case, though, this query will not include any data from the identified intervals.
+        responseContext.put("uncoveredIntervals", uncoveredIntervals);
+        responseContext.put("uncoveredIntervalsOverflowed", uncoveredIntervalsOverflowed);
+      }
+    } else {
+      for (Interval interval : query.getIntervals()) {
+        Iterables.addAll(serversLookup, timeline.lookup(interval));
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/CachingClusteredClient.java]: This is racy between the other `timeline.lookup` call and here. There should only be one `timeline.lookup` call. And if this passes tests then we probably need more concurrency testing for caching clustered client.

Comment (himanshug) [inline server/src/main/java/io/druid/client/CachingClusteredClient.java]: there is only one call ... the code flows are different when uncoveredIntervals are to be populated and not populated. its simpler this way bcoz the code for uncoveredIntervals disabled (else on line 213)  is simple and better on its own.

Comment (drcrallen) [inline server/src/main/java/io/druid/client/CachingClusteredClient.java]: my bad, will look at it closer here.

File: processing/src/main/java/io/druid/query/BaseQuery.java (hunk 1 above)
Comment (himanshug) [inline processing/src/main/java/io/druid/query/BaseQuery.java]: @drcrallen renamed the property to `uncoveredIntervalsLimit`

File: server/src/test/java/io/druid/client/CachingClusteredClientFunctionalityTest.java [hunk 3]
```diff
@@ -81,7 +82,8 @@ public void testUncoveredInterval() throws Exception {
         .dataSource("test")
         .intervals("2015-01-02/2015-01-03")
         .granularity("day")
-        .aggregators(Arrays.<AggregatorFactory>asList(new CountAggregatorFactory("rows")));
+        .aggregators(Arrays.<AggregatorFactory>asList(new CountAggregatorFactory("rows")))
+        .context(ImmutableMap.<String, Object>of("uncoveredIntervalsLimit", 100));
```
Comment (nishantmonu51) [inline server/src/test/java/io/druid/client/CachingClusteredClientFunctionalityTest.java]: are we hitting this limit in the test ? If not can you add a test where we actually hit the limit.

File: server/src/main/java/io/druid/client/CachingClusteredClient.java [hunk 4]
```diff
@@ -166,33 +166,54 @@ public CachingClusteredClient(
...
-      responseContext.put("uncoveredIntervals", uncoveredIntervals);
+      if (!uncoveredIntervals.isEmpty()) {
+        // This returns intervals for which NO segment is present.
+        // Which is not necessarily an indication that the data doesn't exist or is
+        // incomplete. The data could exist and just not be loaded yet.  In either
+        // case, though, this query will not include any data from the identified intervals.
+        responseContext.put("uncoveredIntervals", uncoveredIntervals);
+        responseContext.put("uncoveredIntervalsOverflowed", uncoveredIntervalsOverflowed);
+      }
+    } else {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/CachingClusteredClient.java]: Do tests hit both branches here?

Comment (himanshug) [inline server/src/main/java/io/druid/client/CachingClusteredClient.java]: tests in CachingClusteredClientTest and CachingClusteredClientFunctionalityTest together hit both paths.

File: server/src/test/java/io/druid/client/CachingClusteredClientFunctionalityTest.java (hunk 3 above)
Comment (himanshug) [inline server/src/test/java/io/druid/client/CachingClusteredClientFunctionalityTest.java]: @nishantmonu51 updated test to hit the limit as well.

**Comment by himanshug** [comment]
@nishantmonu51 @drcrallen all comments addressed.

BTW: i don't thin this PR is "Incompatible" , "uncoveredInterval" was not part of 0.8.3 and in 0.9.0 default behavior is same as earlier.

**Comment by CHOIJAEHONG1** [comment]
@himanshug I am just wondering whether there is any particular reason for "uncoveredInterval" to be populated in a HTTP header in place of the HTTP body content?

**Comment by himanshug** [comment]
@CHOIJAEHONG1 because it was originally intended to be for information purposes, in future, I guess we might have a separate section in the response payload for things like this. this PR is only intended to fix the bug really.

**Comment by drcrallen** [comment]
@himanshug this changes default behavior. If people want the behavior of reporting uncovered intervals, then they need to set a new configuration option

**Comment by drcrallen** [comment]
Prior to this patch: return uncovered intervals (or try to)

With this patch: Do not return uncovered intervals

With this patch AND `uncoveredIntervalsLimit` : return uncovered intervals (or try to)

**Comment by drcrallen** [comment]
my bad on the uncovered intervals thing I thought we already had that in 8.3

File: server/src/main/java/io/druid/client/CachingClusteredClient.java (hunk 2 above)
Comment (drcrallen) [inline server/src/main/java/io/druid/client/CachingClusteredClient.java]: timeline.lookup doesn't have a contract about always guaranteeing to return non-null but it does seem to. I'm adding a test for sanity validation of it in another PR.

**Comment by drcrallen** [comment]
:+1:

**Comment by nishantmonu51** [comment]
:+1:

**Pull request merged**